R/ncbi_snp_api.R

```diff
@@ -0,0 +1,322 @@
+         "you must supply rs numbers and they should be prefixed with ",
+         "'rs', e.g. rs420358", call. = FALSE)
+  }
+  
+  message(paste0("Getting info about the following rsIDs: ",
+                 paste(SNPs,
+                       collapse = ", ")))
+  ## transform all SNPs into numbers (rsid)
+  SNPs_num <- gsub("rs", "", SNPs)
+  
+  out <- as.data.frame(matrix(0, nrow = length(SNPs_num), ncol = 15))
+  names(out) <- c("Query", "Chromosome", "BP", "Class", "Marker", "Gene", "Alleles", "AncestralAllele", "VariationAllele", "seqname", "hgvs", "assembly", "ref_seq", "Minor", "MAF")
+  
+  ## as far as I understand from https://api.ncbi.nlm.nih.gov/variation/v0/#/RefSNP/ we
+  ## can only send one query at a time and max 1 per second.
+  for (i in seq_along(SNPs_num)) {
+    
+    variant.url <- paste0("https://api.ncbi.nlm.nih.gov/variation/v0/refsnp/", SNPs_num[i])
+    variant.response <-  httr::GET(variant.url)
+    variant.response.content <-  RJSONIO::fromJSON(rawToChar(variant.response$content),
+                                                   simplifyWithNames = TRUE)
+    
+    if ("error" %in% names(variant.response.content)) {
+      if (variant.response.content$error$message == "RefSNP not found") {
+        warning("The following rsId had no information available on NCBI:\n  ",
+                paste0("rs", SNPs_num[i]),
+                call. = FALSE)
+        next()
+      } else {
+        warning("The following error was received from NCBI:\n  ",
+                variant.response.content$error$message,
+                call. = FALSE)
+        next()
+      }
+    }
+    
+    if ("withdrawn_snapshot_data" %in% names(variant.response.content) & length(variant.response.content$present_obs_movements) == 0) {
+      warning("The following rsId has been withdrawn from NCBI:\n  ",
+              paste0("rs", SNPs_num[i]),
+              call. = FALSE)
+      next()
+    }
+    Query <- as.character(paste0("rs", variant.response.content$refsnp_id))
+    
+    ## if merged into another id
+    if (is.null(variant.response.content$primary_snapshot_data)) {
+      
+      Marker <- as.character(paste0("rs", variant.response.content$merged_snapshot_data$merged_into))
+      no_rs_Marker <-  as.character(variant.response.content$merged_snapshot_data$merged_into)
+      
+      warning(Query, " has been merged into ", Marker, call. = FALSE)
+      
+      variant.url <- paste0("https://api.ncbi.nlm.nih.gov/variation/v0/refsnp/", no_rs_Marker)
+      variant.response <- httr::GET(variant.url)
+      variant.response.content <- RJSONIO::fromJSON(rawToChar(variant.response$content),
+                                                    simplifyWithNames = TRUE)
+      Class <- as.character(variant.response.content$primary_snapshot_data$variant_type)
+      placement_SNP <- get_placements(variant.response.content)
+      
+      if (Class %in% c("snv", "snp", "delins")) {
```

Comment:
Here are repeated `frequency_SNP <- ` elements. Have you thought about reducing this by including `if(Class %in%...) `into the `get_frequency` function?

Reply:
Very nice idea! makes things much more straightforward.